### PR TITLE
Supporting events flows in extensions emulator

### PR DESCRIPTION
**Pull request by huangjeff5**
### Description

This is a feature branch used to hold all extensions emulator features related to events.

### CODE REVIEW

File: src/emulator/functionsEmulator.ts [hunk 1]
```diff
@@ -1119,6 +1156,11 @@ export class FunctionsEmulator implements EmulatorInstance {
       process.env.PUBSUB_EMULATOR_HOST = pubsubHost;
     }
 
+    const eventarcEmulator = this.getEmulatorInfo(Emulators.EVENTARC);
+    if (eventarcEmulator) {
+      process.env.CLOUD_EVENTARC_EMULATOR_HOST = `http://${formatHost(eventarcEmulator)}`;
```
Comment (huangjeff5) [inline src/emulator/functionsEmulator.ts:1161]: Should this be CLOUD_EVENTARC_EMULATOR_HOST or EVENTARC_EMULATOR_HOST?

Comment (apascal07) [reply src/emulator/functionsEmulator.ts]: CLOUD_EVENTARC_EMULATOR_HOST, resolved in sync!

File: src/emulator/eventarcEmulator.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
+import * as _ from "lodash";
```
Comment (joehan) [inline src/emulator/eventarcEmulator.ts:1]: This is unused - looks like it slipped in accidentally?

File: src/emulator/eventarcEmulator.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+import * as express from "express";
```
Comment (joehan) [inline src/emulator/eventarcEmulator.ts:2]: Nit - I prefer to add a newline between imports from external deps (like express) and imports from other files in this project

File: src/emulator/eventarcEmulator.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+import { CloudEvent } from "./events/types";
+import { EmulatorRegistry } from "./registry";
+
+interface CustomEventTrigger {
+  projectId: string;
+  triggerName: string;
+  eventTrigger: EventTrigger;
+}
+
+interface RequestWithRawBody extends express.Request {
```
Comment (joehan) [inline src/emulator/eventarcEmulator.ts:19]: In firebase-functions, we have this exact same type and we call it Request. For consistency, can we just call this Request (or import the firebase-functions type?)

Comment (apascal07) [reply src/emulator/eventarcEmulator.ts:19]: I believe other emulators (e.g. Functions emulator) also call it RequestWithRawBody which is what this is consistent with. What do you think?

Comment (joehan) [reply src/emulator/eventarcEmulator.ts:19]: Ahh, good point - you're right that its better to be consistent within this repo than with firebase-functions. Good call.

File: src/emulator/eventarcEmulator.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+  private logger = EmulatorLogger.forEmulator(Emulators.EVENTARC);
+  private customEvents: { [key: string]: CustomEventTrigger[] } = {};
+
+  constructor(private args: EventarcEmulatorArgs) {}
+
+  createHubServer(): express.Application {
+    const registerTriggerRoute = `/emulator/v1/projects/:project_id/triggers/:trigger_name(*)`;
+    const registerTriggerHandler: express.RequestHandler = (req, res) => {
+      const projectId = req.params.project_id;
```
Comment (joehan) [inline src/emulator/eventarcEmulator.ts:39]: Should we be returning an error if these are missing on the request?

File: src/emulator/eventarcEmulator.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+  async start(): Promise<void> {
+    const { host, port } = this.getInfo();
+    const server = this.createHubServer().listen(port, host);
+    this.destroyServer = createDestroyer(server);
+    return Promise.resolve();
+  }
+
+  async connect(): Promise<void> {
+    // wip. what to do here?
```
Comment (joehan) [inline src/emulator/eventarcEmulator.ts:146]: Probably nothing - in other emulators, we use this to set up any file watchers we need (ie to watch for changes to security rules in Storage/RTDB, or to watch for code changes in functions).

Can just replace this with `return Promise.resolve()`

File: src/emulator/functionsEmulator.ts [hunk 7]
```diff
@@ -271,7 +271,7 @@ export class FunctionsEmulator implements EmulatorInstance {
 
     // The URL for the function that the other emulators (Firestore, etc) use.
     // TODO(abehaskins): Make the other emulators use the route below and remove this.
-    const backgroundFunctionRoute = `/functions/projects/:project_id/triggers/:trigger_name`;
+    const backgroundFunctionRoute = `/functions/projects/:project_id/triggers/:trigger_name(*)`;
```
Comment (joehan) [inline src/emulator/functionsEmulator.ts:274]: why did we need this change"

**Review by joehan — APPROVED** [review]
Some small fixes requested, but LGTM!

File: src/emulator/functionsEmulator.ts (hunk 7 above)
Comment (apascal07) [reply src/emulator/functionsEmulator.ts:274]: trigger_name is of format <function name>-<fully qualified channel name with forward slashes> so it doesn't parse the path correctly unless we specify that the rest of the path should be used for trigger_name.

Comment (apascal07) [reply src/emulator/functionsEmulator.ts:274]: Err, tags got stripped. Meant to say of format "function name" - "fully qualified channel path with forward slashes".

**Pull request merged**